### ISSUE TITLE
[Riot 3] if-else-elseif example fails the first rendering

### ISSUE THREAD

**Issue by cognitom** [riot/examples#40]

**Comment by GianlucaGuarini** [comment]
the `update` and `updated` event will not be triggered on the first mount. Use the `mount` event instead

**Comment by cognitom** [comment]
Cool! Thanks @GianlucaGuarini

**Comment by cognitom** [comment]
This works:

``` html
<then>
  <virtual if={ show }><yield/></virtual>
  <script>
    check()
    this.on('update', check)
    function check () { this.show = !this.parent.skip && this.parent.ok }
  </script>
</then>
```

but this doesn't:

``` html
<then>
  <virtual if={ show }><yield/></virtual>
  <script>
    this.on('mount', check)
    this.on('update', check)
    function check () { this.show = !this.parent.skip && this.parent.ok }
  </script>
</then>
```

I'm not sure why...
@GianlucaGuarini any thought?

**Comment by GianlucaGuarini** [comment]
@cognitom from a quick look the first check does not bind the this scope, and the second does it.

``` html
<my-tag>
<script>
function check() {
   return this
}
var obj = {}
var self = this
var checkObj = check.bind(obj)
console.log(check()) // => window
console.log(checkObj()) // => obj
this.on('mount', check) // => self
</script>
</my-tag>
```

**Comment by cognitom** [comment]
Oops, it was simply my mistake. Thanks @GianlucaGuarini !
Here's working example and fixed in b9cd938245dc755c19e9e1dc70fa45e32502d170

``` html
<then>
  <virtual if={ show }><yield/></virtual>
  <script>
    check () { this.show = !this.parent.skip && this.parent.ok }
    this.on('update', this.check)
    this.check()
  </script>
</then>
```

But `this.on('mount', this.check)` still doesn't work. It looks that the mount event isn't raised. I'm looking into deeper later.

**Comment by cognitom** [comment]
@GianlucaGuarini maybe I see.

If we change something in `mount` event, nothing appear in the dom until we call `this.update()`. Is this a correct behavior?
http://plnkr.co/edit/XhqhtdtgZWDbWOk4bYRV?p=preview

**Comment by GianlucaGuarini** [comment]
@cognitom yes because we have also a `before-mount` event http://plnkr.co/edit/ni5ky7V3USB5A0NJ5C5z?p=preview

**Comment by cognitom** [comment]
@GianlucaGuarini yeah, that's right! Sorry, I forgot `before-mount`.

**Comment by GianlucaGuarini** [comment]
@cognitom basically `before-mount` and calling the function directly on init [are the same thing](https://github.com/riot/riot/blob/next/lib/browser/tag/tag.js#L229-L231) but in this way it's clearer (at least for me) what should happen before and after the mount event

**Comment by cognitom** [comment]
I see, thanks for the detail. Agreed.
OK, the latest code is [here](https://github.com/riot/examples/blob/next/if-else-elseif/if.tag). Much simpler :smile:

tl;dr, my recommended writing would be like this:

``` javascript
// In ES6
const calculate = () => { /* */ }
this
  .on('before-mount', calculate)
  .on('update', calculate)
```

``` javascript
// In ES5
var self = this
self
  .on('before-mount', calculate)
  .on('update', calculate)

function calculate () { /* */ }
```

**Comment by GianlucaGuarini** [comment]
Eventually.. this is also valid es5 syntax:

``` js
function calculate () { /*  */ }
this
  .on('before-mount', calculate)
  .on('update', calculate)
```

**Comment by cognitom** [comment]
Yep. And sometimes invalid in a complex example. That makes beginners mad... sigh.
